File: aether_messenger/windows/AetherQMessenger/Data/SessionManager.cs
using System;
using System.Collections.Concurrent;
using AetherQMessenger.Ffi;

namespace AetherQMessenger.Data;

/// <summary>
/// Bitta kontakt bilan bo'lgan ratchet-seansi — AUTH-AKEM handshake'dan olingan
/// <c>shared_secret</c>dan hosil qilingan <c>session_id</c>/<c>base_seed</c> asosida.
///
/// PoC chegarasi: bu holat faqat jarayon-ichida (in-memory) saqlanadi — ilova qayta
/// ishga tushganda har bir kontakt uchun handshake avtomatik qayta bajariladi.
/// </summary>
public sealed class ContactSession
{
    public required byte[] SessionId { get; init; }
    public required CompressSenderHandle Sender { get; init; }
    public required CompressRecipientHandle Recipient { get; init; }
}

public sealed class SessionManager
{
    private readonly ConcurrentDictionary<string, ContactSession> _sessions = new();
    private readonly ConcurrentDictionary<string, string> _sessionIdIndex = new();

    public ContactSession? SessionFor(string contactUserId) =>
        _sessions.TryGetValue(contactUserId, out var session) ? session : null;

    public bool HasSession(string contactUserId) => _sessions.ContainsKey(contactUserId);

    public string? ContactForSessionId(byte[] sessionId) =>
        _sessionIdIndex.TryGetValue(Convert.ToHexString(sessionId).ToLowerInvariant(), out var contactId) ? contactId : null;

    /// <summary>AUTH-AKEM handshake natijasidan ratchet-seansini o'rnatadi.</summary>
    public ContactSession EstablishSession(string contactUserId, byte[] sharedSecret)
    {
        var baseSeed = AetherCrypto.DeriveRatchetSeed(sharedSecret);
        var sessionId = AetherCrypto.DeriveSessionId(sharedSecret);
        var session = new ContactSession
        {
            SessionId = sessionId,
            Sender = new CompressSenderHandle(sessionId, baseSeed),
            Recipient = new CompressRecipientHandle(sessionId, baseSeed),
        };
        _sessions[contactUserId] = session;
        _sessionIdIndex[Convert.ToHexString(sessionId).ToLowerInvariant()] = contactUserId;
        return session;
    }
}

File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/data/SessionManager.kt
package com.aetherq.messenger.data

import com.aetherq.messenger.crypto.AetherCrypto
import com.aetherq.messenger.crypto.toHex
import uniffi.aether_ffi.CompressRecipientHandle
import uniffi.aether_ffi.CompressSenderHandle
import java.util.concurrent.ConcurrentHashMap

/**
 * Bitta kontakt bilan bo'lgan ratchet-seansi — AUTH-AKEM handshake'dan olingan
 * `shared_secret`dan hosil qilingan `session_id`/`base_seed` asosida.
 *
 * PoC chegarasi: bu holat faqat jarayon-ichida (in-memory) saqlanadi — ilova qayta
 * ishga tushganda har bir kontakt uchun handshake avtomatik qayta bajariladi
 * (`README.md`dagi "Ma'lum cheklovlar" bo'limiga qarang).
 */
class ContactSession(
    val sessionId: ByteArray,
    val sender: CompressSenderHandle,
    val recipient: CompressRecipientHandle,
)

/** Har bir kontakt uchun `ContactSession`ni yaratadi va saqlaydi. */
class SessionManager {
    private val sessions = ConcurrentHashMap<String, ContactSession>()

    // MQTT'da barcha kontaktlar bitta inbox topic'iga yozadi (aetherq/msgr/<ownUserId>) —
    // kelgan ratchet-message qaysi kontaktga tegishli ekanligini frame ichidagi
    // session_id orqali aniqlash uchun teskari indeks.
    private val sessionIdIndex = ConcurrentHashMap<String, String>()

    fun sessionFor(contactUserId: String): ContactSession? = sessions[contactUserId]

    fun hasSession(contactUserId: String): Boolean = sessions.containsKey(contactUserId)

    fun contactForSessionId(sessionId: ByteArray): String? = sessionIdIndex[sessionId.toHex()]

    /** AUTH-AKEM handshake natijasidan ratchet-seansini o'rnatadi — ikkala tomon ham
     * o'zining lokal `shared_secret`i bilan mustaqil chaqiradi va bir xil natijaga keladi. */
    fun establishSession(contactUserId: String, sharedSecret: ByteArray): ContactSession {
        val baseSeed = AetherCrypto.deriveRatchetSeed(sharedSecret)
        val sessionId = AetherCrypto.deriveSessionId(sharedSecret)
        val session = ContactSession(
            sessionId = sessionId,
            sender = AetherCrypto.newCompressSender(sessionId, baseSeed),
            recipient = AetherCrypto.newCompressRecipient(sessionId, baseSeed),
        )
        sessions[contactUserId] = session
        sessionIdIndex[sessionId.toHex()] = contactUserId
        return session
    }

    fun clear(contactUserId: String) {
        sessions.remove(contactUserId)?.let { sessionIdIndex.remove(it.sessionId.toHex()) }
    }
}
